libcloudforensics/providers/kubernetes/workloads.py

```diff
@@ -30,53 +31,209 @@ class K8sWorkload(base.K8sNamespacedResource, metaclass=abc.ABCMeta):
   """
 
   @abc.abstractmethod
-  def GetCoveredPods(self) -> List[base.K8sPod]:
-    """Gets a list of Kubernetes pods covered by this workload.
+  def _PodMatchLabels(self) -> Dict[str, str]:
+    """Gets the key-value pairs that pods belonging to this workload would have.
 
     Returns:
-      List[K8sPod]: A list of pods covered by this workload.
+      Dict[str, str]: The label key-value pairs of this workload's pods.
     """
 
+  @abc.abstractmethod
+  def Read(self) -> Union[client.V1Deployment, client.V1ReplicaSet,]:
+    """Override of abstract method."""  # Narrows down type hint
 
-class K8sDeployment(K8sWorkload):
-  """Class representing a Kubernetes deployment."""
+  @abc.abstractmethod
+  def OrphanPods(self) -> None:
+    """Orphans the pods covered by this workload.
 
-  def Read(self) -> client.V1Deployment:
-    """Override of abstract method."""
-    api = self._Api(client.AppsV1Api)
-    return api.read_namespaced_deployment(self.name, self.namespace)
+    Note that calling this function may entail the deletion of the object
+    upon which this method was called.
+    """
+
+  @abc.abstractmethod
+  def AddTemplateLabels(self, labels: Dict[str, str]) -> None:
+    """Adds labels to the pod template spec of this deployment.
+
+    Args:
+      labels (Dict[str, str]): The labels to be added to the pod template spec.
+    """
 
   def MatchLabels(self) -> Dict[str, str]:
-    """Gets the labels that will be on the pods of this workload.
+    """Gets the label key-value pairs in the matchLabels field.
 
     Returns:
-      Dict[str, str]: The labels that will be on the pods of this workload.
+      Dict[str, str]: The labels in the matchLabels field.
 
     Raises:
       NotImplementedError: If matchExpressions exist, in which case using the
-        matchLabels will be inaccurate.
+          matchLabels will be inaccurate.
     """
     read = self.Read()
+
     if read.spec.selector.match_expressions is not None:
-      raise NotImplementedError('matchExpressions exist, meaning pods matching '
-                                'matchLabels will be inaccurate.')
-    match_labels: Dict[str, str] = read.spec.selector.match_labels
+      raise NotImplementedError(
```

Comment:
Hmm, so in my view, this function will only be used when providing a filter to Kubernetes API (e.g. give me the pods that have these matchLabels). And if there are matchExpressions, I'd rather not let the program continue execution because the results will be erroneous.

 I would say `NotImplementedError` is appropriate because it may be possible to be returning something more sophisticated in the future (like a subclass of `K8sSelector`), that would be covering the matchExpressions.